language/vm/src/file_format.rs

```diff
@@ -487,43 +486,123 @@ impl Signature {
 /// type parameter in the `FunctionHandle` and `StructHandle`.
 pub type TypeParameterIndex = u16;
 
-/// A `Kind` classifies types into sets with rules each set must follow.
+/// A `Constriant` classifies what operations are permitted for a given type
 ///
 /// Currently there are three kinds in Move: `All`, `Resource` and `Copyable`.
+#[repr(u8)]
 #[derive(Debug, Clone, Eq, Copy, Hash, Ord, PartialEq, PartialOrd)]
 #[cfg_attr(any(test, feature = "fuzzing"), derive(Arbitrary))]
-pub enum Kind {
-    /// Represents the super set of all types. The type might actually be a `Resource` or
-    /// `Copyable` A type might be in this set if it is not known to be a `Resource` or
-    /// `Copyable`
-    ///   - This occurs when there is a type parameter with this kind as a constraint
-    All,
-    /// `Resource` types must follow move semantics and various resource safety rules, namely:
-    /// - `Resource` values cannot be copied
-    /// - `Resource` values cannot be popped, i.e. they must be used
-    Resource,
-    /// `Copyable` types do not need to follow the `Resource` rules.
-    /// - `Copyable` values can be copied
-    /// - `Copyable` values can be popped
-    Copyable,
-}
-
-impl Kind {
-    /// Checks if the given kind is a sub-kind of another.
-    #[inline]
-    pub fn is_sub_kind_of(self, k: Kind) -> bool {
-        use Kind::*;
+pub enum Ability {
+    /// Allows values of that type to be copied, via CopyLoc or ReadRef
+    Copy = 0x1,
+    /// Allows values of that type to be dropped, via Pop, WriteRef, StLoc, or if left in a
+    /// local when Ret is invoked
+    Drop = 0x2,
+    /// Allows values of that type to exist inside of a resource in global storage
+    Store = 0x4,
+    /// Allows the type to serve as a key for global storage operations: MoveTo, MoveFrom, etc.
+    Resource = 0x8,
+}
+
+impl Ability {
+    fn from_u8(u: u8) -> Option<Self> {
+        match u {
+            0x1 => Some(Ability::Copy),
+            0x2 => Some(Ability::Drop),
+            0x4 => Some(Ability::Store),
+            0x8 => Some(Ability::Resource),
+            _ => None,
+        }
+    }
+}
+
+/// A set of `Ability`s
```

Comment:
```suggestion
/// A set of `Ability`s encoded as a bitvector
```

Reply:
Just checking: extending this to be (e.g.) `u64` in the future if we add more ability's will be a non-breaking change, right?

Reply:
I think so, since it fits under the first 7 bits, it could be switched to uleb128.
Maybe @matbd or @vgao1996 could double check my work here